File: java/com/epicdragonworld/gameserver/network/ReceivablePacket.java
package com.epicdragonworld.gameserver.network;

import java.io.ByteArrayInputStream;

/**
 * @author Pantelis Andrianakis
 */
public class ReceivablePacket
{
	private final ByteArrayInputStream _bais;
	
	public ReceivablePacket(byte[] bytes)
	{
		_bais = new ByteArrayInputStream(bytes);
	}
	
	public String readString()
	{
		String result = "";
		try
		{
			result = new String(readBytes(readByte()), "UTF-8");
		}
		catch (Exception e)
		{
		}
		return result;
	}
	
	public byte[] readBytes(int length)
	{
		final byte[] result = new byte[length];
		for (int i = 0; i < length; i++)
		{
			result[i] = (byte) _bais.read();
		}
		return result;
	}
	
	public int readByte()
	{
		return _bais.read();
	}
	
	public int readShort()
	{
		return (_bais.read() & 0xff) | ((_bais.read() << 8) & 0xff00);
	}
	
	public int readInt()
	{
		int result = _bais.read() & 0xff;
		result |= (_bais.read() << 8) & 0xff00;
		result |= (_bais.read() << 0x10) & 0xff0000;
		result |= (_bais.read() << 0x18) & 0xff000000;
		return result;
	}
	
	public long readLong()
	{
		long result = _bais.read() & 0xff;
		result |= (_bais.read() & 0xffL) << 8L;
		result |= (_bais.read() & 0xffL) << 16L;
		result |= (_bais.read() & 0xffL) << 24L;
		result |= (_bais.read() & 0xffL) << 32L;
		result |= (_bais.read() & 0xffL) << 40L;
		result |= (_bais.read() & 0xffL) << 48L;
		result |= (_bais.read() & 0xffL) << 56L;
		return result;
	}
	
	public float readFloat()
	{
		int result = _bais.read() & 0xff;
		result |= (_bais.read() << 8) & 0xff00;
		result |= (_bais.read() << 0x10) & 0xff0000;
		result |= (_bais.read() << 0x18) & 0xff000000;
		return Float.intBitsToFloat(result);
	}
	
	public double readDouble()
	{
		long result = _bais.read() & 0xff;
		result |= (_bais.read() & 0xffL) << 8L;
		result |= (_bais.read() & 0xffL) << 16L;
		result |= (_bais.read() & 0xffL) << 24L;
		result |= (_bais.read() & 0xffL) << 32L;
		result |= (_bais.read() & 0xffL) << 40L;
		result |= (_bais.read() & 0xffL) << 48L;
		result |= (_bais.read() & 0xffL) << 56L;
		return Double.longBitsToDouble(result);
	}
}

File: EpicDragonWorldServer/Network/ReceivablePacket.cs
using System;
using System.IO;
using System.Text;

/**
 * Author: Pantelis Andrianakis
 * Date: December 23rd 2017
 */
public class ReceivablePacket
{
    private readonly MemoryStream memoryStream;

    public ReceivablePacket(byte[] bytes)
    {
        memoryStream = new MemoryStream(bytes);
    }

    public string ReadString()
    {
        // Since we use short value maximum byte size for strings is 32767.
        // Take care that maximum packet size data is 32767 bytes as well.
        // Sending a 32767 byte string would require all the available packet size.
        return Encoding.UTF8.GetString(ReadBytes(ReadShort()));
    }

    public byte[] ReadBytes(int length)
    {
        byte[] result = new byte[length];
        for (int i = 0; i < length; i++)
        {
            result[i] = (byte)memoryStream.ReadByte();
        }
        return result;
    }

    public int ReadByte()
    {
        return memoryStream.ReadByte();
    }

    public int ReadShort()
    {
        byte[] byteArray = new byte[2];
        byteArray[0] = (byte)memoryStream.ReadByte();
        byteArray[1] = (byte)memoryStream.ReadByte();
        return BitConverter.ToInt16(byteArray);
    }

    public int ReadInt()
    {
        byte[] byteArray = new byte[4];
        byteArray[0] = (byte)memoryStream.ReadByte();
        byteArray[1] = (byte)memoryStream.ReadByte();
        byteArray[2] = (byte)memoryStream.ReadByte();
        byteArray[3] = (byte)memoryStream.ReadByte();
        return BitConverter.ToInt32(byteArray);
    }

    public long ReadLong()
    {
        byte[] byteArray = new byte[8];
        byteArray[0] = (byte)memoryStream.ReadByte();
        byteArray[1] = (byte)memoryStream.ReadByte();
        byteArray[2] = (byte)memoryStream.ReadByte();
        byteArray[3] = (byte)memoryStream.ReadByte();
        byteArray[4] = (byte)memoryStream.ReadByte();
        byteArray[5] = (byte)memoryStream.ReadByte();
        byteArray[6] = (byte)memoryStream.ReadByte();
        byteArray[7] = (byte)memoryStream.ReadByte();
        return BitConverter.ToInt64(byteArray);
    }

    public float ReadFloat()
    {
        byte[] byteArray = new byte[4];
        byteArray[0] = (byte)memoryStream.ReadByte();
        byteArray[1] = (byte)memoryStream.ReadByte();
        byteArray[2] = (byte)memoryStream.ReadByte();
        byteArray[3] = (byte)memoryStream.ReadByte();
        return BitConverter.ToSingle(byteArray);
    }

    public double ReadDouble()
    {
        byte[] byteArray = new byte[8];
        byteArray[0] = (byte)memoryStream.ReadByte();
        byteArray[1] = (byte)memoryStream.ReadByte();
        byteArray[2] = (byte)memoryStream.ReadByte();
        byteArray[3] = (byte)memoryStream.ReadByte();
        byteArray[4] = (byte)memoryStream.ReadByte();
        byteArray[5] = (byte)memoryStream.ReadByte();
        byteArray[6] = (byte)memoryStream.ReadByte();
        byteArray[7] = (byte)memoryStream.ReadByte();
        return BitConverter.ToDouble(byteArray);
    }
}
